sdk/trace/provider.go

```diff
@@ -278,27 +251,61 @@ func WithSpanProcessor(sp SpanProcessor) TracerProviderOption {
 // The resource is added to the span when it is started.
 func WithResource(r *resource.Resource) TracerProviderOption {
 	return func(opts *TracerProviderConfig) {
-		opts.config.Resource = r
+		if r != nil {
+			opts.resource = r
+		}
 	}
 }
 
 // WithIDGenerator option registers an IDGenerator with the TracerProvider.
 func WithIDGenerator(g IDGenerator) TracerProviderOption {
 	return func(opts *TracerProviderConfig) {
-		opts.config.IDGenerator = g
+		if g != nil {
+			opts.idGenerator = g
+		}
 	}
 }
 
 // WithDefaultSampler option registers a DefaultSampler with the the TracerProvider.
 func WithDefaultSampler(s Sampler) TracerProviderOption {
 	return func(opts *TracerProviderConfig) {
-		opts.config.DefaultSampler = s
+		if s != nil {
+			opts.sampler = s
+		}
 	}
 }
 
 // WithSpanLimits option registers a SpanLimits with the the TracerProvider.
 func WithSpanLimits(sl SpanLimits) TracerProviderOption {
 	return func(opts *TracerProviderConfig) {
-		opts.config.SpanLimits = sl
+		opts.spanLimits = sl
+	}
+}
+
+// ensureValidTracerProviderConfig ensures that given TracerProviderConfig is valid.
+func ensureValidTracerProviderConfig(cfg *TracerProviderConfig) {
+	if cfg.sampler == nil {
+		cfg.sampler = ParentBased(AlwaysSample())
+	}
+	if cfg.idGenerator == nil {
+		cfg.idGenerator = defaultIDGenerator()
+	}
+	if cfg.spanLimits.EventCountLimit <= 0 {
+		cfg.spanLimits.EventCountLimit = DefaultEventCountLimit
+	}
+	if cfg.spanLimits.AttributeCountLimit <= 0 {
+		cfg.spanLimits.AttributeCountLimit = DefaultAttributeCountLimit
+	}
+	if cfg.spanLimits.LinkCountLimit <= 0 {
+		cfg.spanLimits.LinkCountLimit = DefaultLinkCountLimit
+	}
+	if cfg.spanLimits.AttributePerEventCountLimit <= 0 {
+		cfg.spanLimits.AttributePerEventCountLimit = DefaultAttributePerEventCountLimit
+	}
+	if cfg.spanLimits.AttributePerLinkCountLimit <= 0 {
+		cfg.spanLimits.AttributePerLinkCountLimit = DefaultAttributePerLinkCountLimit
+	}
```

Comment:
Move validation of span limits to its own function.

Reply:
https://github.com/open-telemetry/opentelemetry-go/issues/1705

Reply:
Done.

Reply:
Sorry, meant for this to be tackled in separate PR.